From 45bb5e06b71760c9499dad40f2e8a2c7bfa41fda Mon Sep 17 00:00:00 2001
From: Holger Weiss <holger@zedat.fu-berlin.de>
Date: Sat, 30 Dec 2017 01:14:07 +0100
Subject: [PATCH] Fix a memory leak in the get_certfile/1 NIF

---
 c_src/fast_tls.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/c_src/fast_tls.c b/c_src/fast_tls.c
index ba43f76..efacb7c 100644
--- a/c_src/fast_tls.c
+++ b/c_src/fast_tls.c
@@ -1316,6 +1316,7 @@ static ERL_NIF_TERM get_certfile_nif(ErlNifEnv *env, int argc,
       result = enif_make_atom(env, "error");
     }
     hashmap_unlock(certfiles_map, 0);
+    enif_free(key);
   } else {
     result = enif_make_atom(env, "error");
   }
